RecoLocalTracker/SiPixelRecHits/src/PixelCPEFast.cc

```diff
@@ -185,21 +192,103 @@ void PixelCPEFast::fillParamsForGpu() {
+                               << ' ' << int(g.sigmax1[ix]) << ' ' << 10000.f * cp.sigmay << std::endl;
+#endif  // EDM_ML_DEBUG
+    }
+#ifdef EDM_ML_DEBUG
+    // sample yerr as function of position
+    auto yoff = -54 * 4.f * commonParamsGPU_.thePitchY;
+    for (int ix = 0; ix < 16; ++ix) {
+      auto y = yoff + (0.5f + float(ix)) * 432.f * commonParamsGPU_.thePitchY / 16.f;
+      auto gvx = p.theOrigin.x() + 40.f * commonParamsGPU_.thePitchY;
+      auto gvy = p.theOrigin.y() - y;
+      auto gvz = 1.f / p.theOrigin.z();
+      cp.cotbeta = gvy * gvz;
+      cp.cotalpha = gvx * gvz;
+      errorFromTemplates(p, cp, 20000.f);
+      LogDebug("PixelCPEFast") << "sigmay vs y " << i << ' ' << y << ' ' << cp.cotbeta << ' ' << 10000.f * cp.sigmay
+                               << std::endl;
+    }
+#endif  // EDM_ML_DEBUG
+
+    // average angle
+    auto gvx = p.theOrigin.x() + 40.f * commonParamsGPU_.thePitchX;
+    auto gvy = p.theOrigin.y();
+    auto gvz = 1.f / p.theOrigin.z();
+    //--- Note that the normalization is not required as only the ratio used
 
-    g.sy[0] = cp.sigmay;
-    g.sy[1] = cp.sy1;
-    g.sy[2] = cp.sy2;
+    // calculate angles
+    cp.cotalpha = gvx * gvz;
+    cp.cotbeta = gvy * gvz;
+    auto aveCB = cp.cotbeta;
 
-    for (int i = 0; i < 3; ++i) {
-      g.sx[i] = std::sqrt(g.sx[i] * g.sx[i] + lape.xx());
-      g.sy[i] = std::sqrt(g.sy[i] * g.sy[i] + lape.yy());
+    // sample x by charge
+    int qbin = 5;  // low charge
+    int k = 0;
+    for (int qclus = 1000; qclus < 200000; qclus += 1000) {
+      errorFromTemplates(p, cp, qclus);
+      if (cp.qBin_ == qbin)
+        continue;
+      qbin = cp.qBin_;
+      g.xfact[k] = cp.sigmax;
+      g.yfact[k] = cp.sigmay;
+      g.minCh[k++] = qclus;
+#ifdef EDM_ML_DEBUG
+      LogDebug("PixelCPEFast") << i << ' ' << g.rawId << ' ' << cp.cotalpha << ' ' << qclus << ' ' << cp.qBin_ << ' '
+                               << cp.pixmx << ' ' << m * cp.sigmax << ' ' << m * cp.sx1 << ' ' << m * cp.sx2 << ' '
+                               << m * cp.sigmay << ' ' << m * cp.sy1 << ' ' << m * cp.sy2 << std::endl;
+#endif  // EDM_ML_DEBUG
     }
-  }
+    assert(k <= 5);
+    // fill the rest  (sometimes bin 4 is missing)
+    for (int kk = k; kk < 5; ++kk) {
+      g.xfact[kk] = g.xfact[k - 1];
+      g.yfact[kk] = g.yfact[k - 1];
+      g.minCh[kk] = g.minCh[k - 1];
+    }
+    auto detx = 1.f / g.xfact[0];
+    auto dety = 1.f / g.yfact[0];
+    for (int kk = 0; kk < 5; ++kk) {
+      g.xfact[kk] *= detx;
+      g.yfact[kk] *= dety;
+    }
+    // sample y in angle
+    float ys = 8.f - 4.f;  // apperent bias of half pixel (see plot)
+    // sample yerr as function of "size"
+    for (int iy = 0; iy < 16; ++iy) {
+      ys += 1.f;  // first bin 0 is for size 9  (and size is in fixed point 2^3)
+      if (15 == iy)
+        ys += 8.f;  // last bin for "overflow"
+      // cp.cotalpha = ys*100.f/(8.f*285.f);
```

Comment:
is this commented out code needed?
Please remove or add comments inline in the code why the commented out block is relevant